lib/tasks/integration.rake

```diff
@@ -1,144 +1,218 @@
-  # rake doc:swagger:validate:all in 6.3s
-  # rake ci:lint --trace in 7.4s
-  # rake ci:jspm --trace in 15.2s
-  # yarn test -- --reporters dots,junit --browsers Firefox in 10.7s
-  # rake db:setup in 10.3s
-
-  # parallel_cucumber --verbose features -o "-b -p parallel --tags=~@javascript --tags=~@backend --tags=~@emails --tags=~@stats --tags=~@search --tags=~@no-txn" in 428.6s
-
-  # parallel_cucumber --verbose features -o "-b -p parallel --tags=~@javascript --tags=@backend,@emails,@stats,@search,@no-txn" in 410.2s
-
-  kind = {
-    '1' => [
-      'parallel_cucumber --verbose features -o "-b -p parallel --tags=@javascript --tags=~@fakeweb --tags=~@percy"',
-      'parallel_rspec --verbose spec',
-    ],
-    '2' => [
-      "parallel_test --verbose #{test_dirs.delete('test/integration')}",
-    ],
-    '3' => [
+    :rspec => 'parallel_rspec --verbose spec',
+    :integration => "parallel_test --verbose #{test_dirs.delete('test/integration')}",
+    :swagger => [
       'rake doc:swagger:validate:all',
       'rake doc:swagger:generate:all',
+    ],
+    :frontend => [
       'rake ci:jspm --trace',
       'yarn test -- --reporters dots,junit --browsers Firefox',
       'yarn jest',
       'rake db:purge db:setup',
     ],
-    '4' => [
-      "parallel_test --verbose #{test_dirs.delete('test/functional')}",
-      "parallel_test --verbose #{test_dirs.join(' ')}",
-    ],
-    '5' => [
-      %Q{parallel_cucumber --verbose features -o "-b -p parallel --tags=~@javascript #{tags.map{|t| %Q|--tags=~#{t}| }.join(' ')}"},
-    ],
-    '6' => [
-      %Q{parallel_cucumber --verbose features -o "-b -p parallel --tags=~@javascript --tags=#{tags.join(',')}"},
-    ],
-
-    'percy' => [
-      'PERCY_ENABLE=1 cucumber -b -p parallel --tags=@percy features'
-    ],
-    'licenses' => [
-      "export http_proxy=#{ENV['http_proxy']} https_proxy=#{ENV['https_proxy']}; rake ci:license_finder:run"
-    ]
+    :functional => "parallel_test --verbose #{test_dirs.delete('test/functional')}",
+    :license_checks => "export http_proxy=#{ENV['http_proxy']} https_proxy=#{ENV['https_proxy']}; rake ci:license_finder:run",
+    :main_suite => "parallel_test --verbose #{test_dirs.join(' ')}",
+    :percy => 'PERCY_ENABLE=1 cucumber -b -p parallel --tags=@percy features',
   }
 
-  tasks = ENV['MULTIJOB_KIND'].present? ? kind.fetch(ENV['MULTIJOB_KIND']) : kind.values.flatten
+end
+# rubocop:enable MethodLength
 
-  success, failure = "#{Color::GREEN}SUCCESS#{Color::CLEAR_COLOR}", "#{Color::RED}FAILURE#{Color::CLEAR_COLOR}"
 
-  summary = []
 
-  banner = ->(line) do
-    puts
-    puts "=" * 40
-    puts "== #{Color::BOLD}#{line}#{Color::CLEAR}"
-    puts "=" * 40
-    puts
 
-    line
+desc 'The default execution: the whole CI suite'
+task :integrate => 'integrate:parallel'
+
+
+namespace :integrate do
+
+  desc 'Runs the set of tests passed as an argument, pretty formatting results, creating reports, etc.'
+  task :run_tests, [:some_tests] do |t, args|
```

Comment:
`some_tests` is not a great name. It is actually a command, right?

Reply:
Also, why this should be a task? Can it be executed by a developer?

I think it is more of an internal helper, so it could be just a method or a lambda.

Reply:
agreed about `some_tests`

on 2nd point, after coming back with a 2nd thought on it, I would like to keep this as a task, yes, exactly because it allows the developer to run any arbitrary set of tests by tuning the command itself (which is passed as an argument). 

This seems to be useful because the way we've split tests now seems pretty specific... I could envisage the need for developers to want to tune that (especially when working on specific features)... 

On this basis, I would suggest we do keep it as a task.

Reply:
I don't see how you'd run this task from a command line and pass it arguments.
When you give it description it shows up on a list of available tasks when running `rake -T`. 
Is this "public API" that developers are supposed to use?
Why would a developer need to wrap any arbitrary command into this instead of just running the command? 

Developers can still tune how we split tests. Because that will generate other tasks, that use this. But I don't really see who will ever need to run this from the CLI. 

But it is your call anyway ;)

Reply:
yes, i see what you mean - perhaps for the `dev` environment it _is_ kind of pointless the way the method is now, because there's not much else happening around the command itself that is being executed. 

I was thinking that I would like offering this wrapper around the `command` as part of the API, because it gives us a place to group common tasks that we would always want to happen before/after running the `command` (e.g. think invoking `clean` before `test`). 

There is also the case of the test/CI environment, where I think this makes a little more sense. Can i run any set of tests I choose to without changing the app source code, but only the CI server config?

(Not sure if this will make much sense, but) I see these `commands` as a `compute` workload that is being assigned to run in some `place`. As the orchestrator of that `workload`, I would like flexibility around being able to assign that to different `places` as well as to modify the `workload` itself. If the `workload` definitions live only inside the `rake` files, I am somewhat limited as to what `workloads` I can choose to run (those that the application developers thought I might need to... but as the orchestrator, I might actually know better ;)  )

Reply:
@gsaslis every command that is going to be executed by this should already be able to clean after itself. You are free as a developer to run `cucumber features/foo.feature` and it has to just work.

I don't see this command is used in CI/test environment directly. It is being used by other tasks, yes, but not directly by the CI setup. 

The only thing this task does is that it starts a command you have it, prints a banner around it, waits for it to finish and writes a junit result. It is used to separate nicely the failures in the log and provide JUnit info about the commands executed.

I just don't see where else other than this Rake you want to use: `rake integrate:run_tests[cucumber features/finance/*.feature]` for example. This is always defined in Rake itself, so it does not have to be exposed to the public. The moment you need to use it externally, expose it.

Reply:
yes, i think we both agree that there is a need for someone to run `cucumber features/finance/*.feature` as a workload, while someone else would want to run `cucumber features/something_else/*.feature`. In fact, I think this is so _both_ in `CI/test` as well as in `dev` envs. (e.g. the CI might want to decide to run a different set of tests next time round, because it wants to schedule it differently -- assuming it is smart enough to understand that).

Now, all I'm saying, in addition to the above, is that a wrapper around that (which, yes, at the moment just prints a banner around it and does 1-2 more things) is nice to have as well. 

if `rake` is our build tool, why not offer a way for folks to run their own tests through `rake`, as part of its API, rather than around it? Perhaps it's my non-ruby background that's causing my bias though here (maven, etc.)... :)

Reply:
I think the CI should have no notion of what tests it is running. It should just run one command that runs all the tests. If there is some need to split the workload, it should be defined in Rake, not in the CI/test scripts. All that so there is less opportunity to miss something.

If rake is our build tool, then this can be internal to rake. Not exposed to the CLI. It can be a method, a lambda or whatever that is not exposed to the CLI and it can still be used from rake.

By exposing this you are making it usable from outside of rake and I believe there should be no setup or decisions what test suite to run taken outside of rake.

Reply:
well, the scenario I have in mind is the one where we want to somehow parallelize "all the tests" that the CI should run. 

It would be pretty desirable for me to have the CI to (optimally) split "all the tests" and schedule them accordingly in different `places` by itself. How this `workload` will be scheduled is a decision that I think should live, _not_ inside the app itself, but in the CI (or somewhere else?) 

if the above is a valid use case, then the CI would need some way to set this all up, and - the way I've envisioned this (but I could still be missing sth) - that way would be through the `run_tests` task

Reply:
I can see 3 points here.

1) My point that unless you need it right now, then don't do it is still valid. If it is not needed to be exposed, then don't and do it later when it is actually needed.

2) How to split the test suite. There are two approaches: push (schedule) and pull (queue). 
Either we schedule all the work to a queue that is later processed by N workers and we don't have to care about how many N is or how they are scaled. Or we know how many nodes there are and can schedule work on each of those nodes taking into account time it took the previous run. If we take the queue approach, there is no need for this command to be exposed to the CLI. All tasks would be predefined by rake. If we take the schedule approach, then it is also not needed because we can decide the list of tests to run inside rake. For example, shelling out to the `circleci` command to split the suite based on previous runs. Just run one rake task, that discovers what tests to run based on the node index and how many nodes there are (already done by the circleci CLI).

3) How to define where to run what. We have. N manually configured groups we can run in parallel. If we add more nodes, it would not help parallelization unless we do point 2).
In our case we can still define the rake tasks based on the node index runs the appropriate group of tests. That can still be done in Rake.

My point is: let's try not to predict feature (this might be useful). It either is useful right now, or it isn't. If it becomes useful in the future, then use it.

Reply:
ok, ok, i _do_ accept the YAGNI point :) ;) 

i guess we'll come back to this when we do have a solution for 2/3 (hopefully won't be too long before then)